File: coffeetime-parent/coffeetime-scala/src/main/scala/org/mrseasons/coffeetime/scala/_12_class/Person.scala
package org.mrseasons.coffeetime.scala._12_class

import java.util.Date

/**
 * Created by mrseasons on 2/12/15.
 */
class Person {
  //only getter
  val birthday = new Date()
  //default getter and setter
  var age = 0
  //specify getter and setter
  private var privateAge = 0
  //object private fields
  private[this] var selfAge = 0

  def trueAge = privateAge

  def trueAge_=(pAge: Int): Unit = {
    if (pAge > privateAge) {
      privateAge = pAge
    }
  }

  def isYounger(other: Person) = age < other.age

  //error, is inaccessible
  //  def isYounger2(other:Person)=selfAge<other.selfAge

}

File: coffeetime-parent/coffeetime-kotlin/src/main/kotlin/org/mrseasons/coffeetime/kotlin/_11_class/Person.kt
package org.mrseasons.coffeetime.kotlin._11_class

import java.util.Date

/**
 * Created by mrseasons on 2015/06/05.
 */
public class Person {
    //  default getter and setter
    var address = ""
    public var age: Int = 0

    //  specify getter and setter
    var trueAge: Int
        get() = age
        set(pAge) {
            if (pAge > age) {
                age = pAge
            }
        }

    //  backing field should start with "$"
    //  must be initialized first
    //  if not using "$", backAge will call the getter and stack overflow may occurs
    var backAge: Int = 0
        get() = $backAge
        set(pAge) {
            if (pAge > $backAge) {
                $backAge = pAge
            }
        }

    // change the visiblity
    var setterVisiblity: String = "abc"
        private set

    //  only getter
    val birthday = Date()
}

//  class with no body
class Empty

